crates/unjson/src/unjson.rs

```diff
@@ -73,36 +73,36 @@ impl ValueEx for Value {
 
     fn as_i64_opt(&self) -> Result<Option<i64>> {
         Ok(match self {
-            &Value::Null       => None,
-            &Value::I64(ref v) => Some(*v),
+            &Value::Null          => None,
+            &Value::Number(ref v) if v.is_i64() => Some(v.as_i64().unwrap()),
             _ => { return type_error("i64 or null", self.ty()); }
         })
     }
 
     fn as_u64_opt(&self) -> Result<Option<u64>> {
         Ok(match self {
             &Value::Null       => None,
-            &Value::U64(ref v) => Some(*v),
+            &Value::Number(ref v) if v.is_u64() => Some(v.as_u64().unwrap()),
             _ => { return type_error("u64 or null", self.ty()); }
         })
     }
 
     fn as_f64_opt(&self) -> Result<Option<f64>> {
         Ok(match self {
             &Value::Null       => None,
-            &Value::F64(ref v) => Some(*v),
+            &Value::Number(ref v) if v.is_f64() => Some(v.as_f64().unwrap()),
             _ => { return type_error("f64 or null", self.ty()); }
         })
     }
 
     fn as_number_opt(&self) -> Result<Option<f64>> {
-        Ok(match self {
-            &Value::Null       => None,
-            &Value::I64(ref v) => Some(*v as f64),
-            &Value::U64(ref v) => Some(*v as f64),
-            &Value::F64(ref v) => Some(*v),
-            _ => { return type_error("number or null", self.ty()); }
-        })
+        match *self {
+            Value::Number(ref v) if v.is_f64() => Ok(Some(v.as_f64().unwrap())),
```

Comment:
Well, in the current version of serde_json, `as_f64()` returns `None` if the number is `PosInt` or `NegInt`.

As a side-note, I don't understand *why* they have `PosInt` or `NegInt` in the first place, since these don't exist in JS, nor why `as_f64()` doesn't attempt to convert.